File: lib/core/validation/validation_result.dart
import 'package:nimon/core/validation/validation_issue.dart';
import 'package:nimon/core/validation/validation_severity.dart';

class ValidationResult {
  const ValidationResult({required this.ok, required this.issues});

  final bool ok;
  final List<ValidationIssue> issues;
}

ValidationResult okResult() => const ValidationResult(ok: true, issues: []);

ValidationResult failResult(ValidationIssue issue) => ValidationResult(
      ok: issue.severity != ValidationSeverity.blocking,
      issues: [issue],
    );

ValidationResult combineResults(Iterable<ValidationResult> results) {
  final issues = results.expand((r) => r.issues).toList();
  final blocking = issues.any((i) => i.severity == ValidationSeverity.blocking);
  return ValidationResult(ok: !blocking, issues: issues);
}

ValidationResult resultFromIssues(List<ValidationIssue> issues) {
  final blocking = issues.any((i) => i.severity == ValidationSeverity.blocking);
  return ValidationResult(ok: !blocking, issues: issues);
}

bool hasBlockingIssues(ValidationResult r) =>
    r.issues.any((i) => i.severity == ValidationSeverity.blocking);

File: nimon-backend/src/common/validation/validation-result.ts
import type { ValidationIssue } from './validation-issue';
import { ValidationSeverity } from './validation-severity';

export type ValidationResult = {
  ok: boolean;
  issues: ValidationIssue[];
};

export function ok(): ValidationResult {
  return { ok: true, issues: [] };
}

export function fail(issue: ValidationIssue): ValidationResult {
  return {
    ok: issue.severity !== ValidationSeverity.Blocking,
    issues: [issue],
  };
}

/** Merges multiple results; ok is true only when no blocking issues in any part. */
export function combine(...results: ValidationResult[]): ValidationResult {
  const issues = results.flatMap((r) => r.issues);
  const blocking = issues.some((i) => i.severity === ValidationSeverity.Blocking);
  return { ok: !blocking, issues };
}

export function hasBlockingIssues(result: ValidationResult): boolean {
  return result.issues.some((i) => i.severity === ValidationSeverity.Blocking);
}

export function resultFromIssues(issues: ValidationIssue[]): ValidationResult {
  const blocking = issues.some((i) => i.severity === ValidationSeverity.Blocking);
  return { ok: !blocking, issues };
}
